File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
bool iscycledfs(unordered_map<int,vector<int>>&adj,int u,vector<bool>&visited,vector<bool>&inrecursion){
    visited[u]=true;
    inrecursion[u]=true;
    for(int &v : adj[u]){
        if(!visited[v] && iscycledfs(adj,v,visited,inrecursion))
            return true;
        else if(inrecursion[v]==true)
            return true;
    }
    inrecursion[u]=false;
    return false;
}
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int,vector<int>>adj;
        vector<bool>visited(numCourses,false);
        vector<bool>inrecursion(numCourses,false);
        for(auto &vec : prerequisites){
            int a = vec[0];
            int b = vec[1];
            //b-->a
            adj[b].push_back(a);
        }
        for(int i=0;i<numCourses;i++){
            if(!visited[i] && iscycledfs(adj,i,visited,inrecursion))
                return false;
        }
        return true;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def dfs(self, node, adj, vis, pathvis):
        vis[node] = True
        pathvis[node] = True

        for adjnode in adj[node]:
            if not vis[adjnode]:
                if self.dfs(adjnode, adj, vis, pathvis):
                    return True
            elif pathvis[adjnode]:
                return True

        pathvis[node] = False
        return False

    def canFinish(self, numCourses, prerequisites):

        # Build adjacency list
        adj = [[] for _ in range(numCourses)]

        for a, b in prerequisites:
            adj[b].append(a)

        vis = [False] * numCourses
        pathvis = [False] * numCourses

        # Run DFS from every unvisited node
        for i in range(numCourses):
            if not vis[i]:
                if self.dfs(i, adj, vis, pathvis):
                    return False

        return True